zerver/tests/test_thumbnail.py

```diff
@@ -45,3 +47,53 @@ def test_thumbnail_redirect(self) -> None:
         self.assertEqual(result.status_code, 302, result)
         base = "https://external-content.zulipcdn.net/external_content/676530cf4b101d56f56cc4a37c6ef4d4fd9b0c03/2f2f7777772e676f6f676c652e636f6d2f696d616765732f737270722f6c6f676f34772e706e67"
         self.assertEqual(base, result.url)
+
+    @override_settings(RATE_LIMITING=True)
+    def test_thumbnail_redirect_for_spectator(self) -> None:
+        self.login("hamlet")
+        fp = StringIO("zulip!")
+        fp.name = "zulip.jpeg"
+
+        result = self.client_post("/json/user_uploads", {"file": fp})
+        self.assert_json_success(result)
+        json = orjson.loads(result.content)
+        uri = json["uri"]
+
+        add_ratelimit_rule(86400, 1000, domain="spectator_attachment_access_by_file")
+        # Deny file access for non web public stream
+        self.subscribe(self.example_user("hamlet"), "Denmark")
+        host = self.example_user("hamlet").realm.host
+        body = f"First message ...[zulip.txt](http://{host}" + uri + ")"
+        self.send_stream_message(self.example_user("hamlet"), "Denmark", body, "test")
+
+        self.logout()
+        response = self.client_get("/thumbnail", {"url": uri[1:], "size": "full"})
+        self.assertEqual(response.status_code, 403)
+
+        # Allow file access for web public stream
+        self.login("hamlet")
+        self.make_stream("web-public-stream", is_web_public=True)
+        self.subscribe(self.example_user("hamlet"), "web-public-stream")
+        body = f"First message ...[zulip.txt](http://{host}" + uri + ")"
+        self.send_stream_message(self.example_user("hamlet"), "web-public-stream", body, "test")
+
+        self.logout()
+        response = self.client_get("/thumbnail", {"url": uri[1:], "size": "full"})
+        self.assertEqual(response.status_code, 302)
+        remove_ratelimit_rule(86400, 1000, domain="spectator_attachment_access_by_file")
+
+        # Deny file access since rate limited
+        add_ratelimit_rule(86400, 0, domain="spectator_attachment_access_by_file")
+        response = self.client_get("/thumbnail", {"url": uri[1:], "size": "full"})
+        self.assertEqual(response.status_code, 403)
+        remove_ratelimit_rule(86400, 0, domain="spectator_attachment_access_by_file")
+
+        # Deny random file access
+        response = self.client_get(
+            "/thumbnail",
+            {
+                "url": "user_uploads/2/71/QYB7LA-ULMYEad-QfLMxmI2e/zulip-non-existent.txt",
+                "size": "full",
+            },
+        )
+        self.assertEqual(response.status_code, 403)
```

Comment:
Have you considered writing the rate limiting tests in `test_external.py`, where they perhaps could be written using the `do_test_hit_ratelimits` convention that other tests there use?

Because here the "hit rate limits" case is tested by a adding a special rule allowing 0 requests, while there actually incrementally hitting the limits of a more realistc rule is tested - which be more "comprehensive" testing

Reply:
Oh, didn't know! Since the uploaded file test cases for images were here, I added them here. Also, it looks like I will have to modify `do_test_hit_ratelimits` for it to work in my case.